RecoEcal/EgammaCoreTools/interface/EcalClustersGraph.h

```diff
@@ -0,0 +1,133 @@
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterTools.h"
+#include "DataFormats/EcalDetId/interface/EBDetId.h"
+#include "DataFormats/EcalDetId/interface/EEDetId.h"
+
+#include "Geometry/CaloTopology/interface/CaloTopology.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloSubdetectorGeometry.h"
+#include "Geometry/Records/interface/CaloTopologyRecord.h"
+#include "Geometry/CaloTopology/interface/CaloSubdetectorTopology.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "Geometry/EcalAlgo/interface/EcalBarrelGeometry.h"
+#include "Geometry/EcalAlgo/interface/EcalEndcapGeometry.h"
+
+#include "RecoEcal/EgammaCoreTools/interface/CalibratedPFCluster.h"
+#include "RecoEcal/EgammaCoreTools/interface/GraphMap.h"
+#include "RecoEcal/EgammaCoreTools/interface/SCProducerCache.h"
+
+/*
+ * class:  EcalClustersGraph 
+ * Authors:  D.Valsecchi, B.Marzocchi
+ * Date:  January 2022
+ * 
+ * Utility class to handle all the PFClusters in ECAL as a graph.
+ * The DeepSC algorithm is applied on sub-graphs of clusters to form SuperCluster.
+ */
+
+using namespace std;
+using namespace reco;
+namespace reco {
+
+  class EcalClustersGraph {
+    typedef std::shared_ptr<CalibratedPFCluster> CalibratedClusterPtr;
+    typedef std::vector<CalibratedClusterPtr> CalibratedClusterPtrVector;
+
+  private:
+    CalibratedClusterPtrVector clusters_;
+    uint nSeeds_;
+    uint nCls_;
+
+    //To compute the input variables
+    const CaloTopology* topology_;
+    const CaloSubdetectorGeometry* ebGeom_;
+    const CaloSubdetectorGeometry* eeGeom_;
+    const EcalRecHitCollection* recHitsEB_;
+    const EcalRecHitCollection* recHitsEE_;
+    const SCProducerCache* SCProducerCache_;
+
+    // GraphMap for handling all the windows and scores
+    GraphMap graphMap_;
+    GraphMap::CollectionStrategy strategy_;
+    float threshold_;
+    DeepSCInputs inputs_;
+
+    std::vector<std::pair<uint, std::vector<uint>>> finalSuperClusters_;
+    std::array<float, 3> locCov_;
+    std::pair<double, double> widths_;
+
+  public:
+    EcalClustersGraph(CalibratedClusterPtrVector clusters,
+                      int nSeeds,
+                      const CaloTopology* topology,
+                      const CaloSubdetectorGeometry* ebGeom,
+                      const CaloSubdetectorGeometry* eeGeom,
+                      const EcalRecHitCollection* recHitsEB,
+                      const EcalRecHitCollection* recHitsEE,
+                      const SCProducerCache* cache);
+
+    std::vector<int> clusterPosition(const CaloCluster* cluster);
+
+    double deltaPhi(double seed_phi, double cluster_phi) {
+      double dphi = seed_phi - cluster_phi;
+      if (dphi > TMath::Pi())
+        dphi -= 2 * TMath::Pi();
+      if (dphi < -TMath::Pi())
+        dphi += 2 * TMath::Pi();
+      return dphi;
+    }
+
+    double deltaEta(double seed_eta, double cluster_eta) {
```

Comment:
The convention for deltaEta, also for the Mustache SuperClusters, is to have this sign flip for negative seed eta, so that the "curvature" of the Mustache parabolas is always positive. 

https://github.com/cms-sw/cmssw/blob/master/RecoEcal/EgammaCoreTools/src/Mustache.cc#L60